go/libraries/doltcore/sqle/binlogreplication/binlog_replica_applier.go

```diff
@@ -0,0 +1,997 @@
+
+			foreignKeyChecksDisabled := tableMap.Flags&rowFlag_noForeignKeyChecks > 0
+			writeSession, tableWriter, err := getTableWriter(ctx, engine, tableMap.Name, tableMap.Database, foreignKeyChecksDisabled)
+			if err != nil {
+				return err
+			}
+
+			err = tableWriter.Update(ctx, identifyRow, updatedRow)
+			if err != nil {
+				return err
+			}
+
+			err = closeWriteSession(ctx, engine, tableMap.Database, writeSession)
+			if err != nil {
+				return err
+			}
+		}
+
+	default:
+		// We can't access the bytes directly because these non-interface types in Vitess are not exposed.
+		// Having a Bytes() or Type() method on the Vitess interface would let us clean this up.
+		byteString := fmt.Sprintf("%v", event)
+		if strings.HasPrefix(byteString, "{[0 0 0 0 27 ") {
+			// Type 27 is a Heartbeat event. This event does not appear in the binary log. It's only sent over the
+			// network by a primary to a replica to let it know that the primary is still alive, and is only sent
+			// when the primary has no binlog events to send to replica servers.
+			// For more details, see: https://mariadb.com/kb/en/heartbeat_log_event/
+			logger.Debug("Received binlog event: Heartbeat")
+		} else {
+			return fmt.Errorf("received unknown event: %v", event)
+		}
+	}
+
+	if createCommit {
+		databasesToCommit := keys(a.modifiedDatabases)
+		if commitToAllDatabases {
+			databasesToCommit = getAllUserDatabaseNames(ctx, engine)
+		}
+		for _, database := range databasesToCommit {
+			executeQueryWithEngine(ctx, engine, "use `"+database+"`;")
+			executeQueryWithEngine(ctx, engine, "commit;")
+		}
+
+		// Record the last GTID processed after the commit
+		a.currentPosition.GTIDSet = a.currentPosition.GTIDSet.AddGTID(a.currentGtid)
+		err := sql.SystemVariables.AssignValues(map[string]interface{}{"gtid_executed": a.currentPosition.GTIDSet.String()})
+		if err != nil {
+			logger.Errorf("unable to set @@GLOBAL.gtid_executed: %s", err.Error())
+		}
+		err = positionStore.Save(ctx, a.currentPosition)
+		if err != nil {
+			return fmt.Errorf("unable to store GTID executed metadata to disk: %s", err.Error())
+		}
+
+		// For now, create a Dolt commit from every data update. Eventually, we'll want to make this configurable.
+		logger.Trace("Creating Dolt commit(s)")
+		for _, database := range databasesToCommit {
+			executeQueryWithEngine(ctx, engine, "use `"+database+"`;")
+			executeQueryWithEngine(ctx, engine,
+				fmt.Sprintf("call dolt_commit('-Am', 'Dolt binlog replica commit: GTID %s');", a.currentGtid))
+		}
+
+		// Clear the modified database metadata for the next commit
+		a.modifiedDatabases = make(map[string]struct{})
+		commitToAllDatabases = false
+	}
+
+	return nil
+}
+
+//
+// Helper functions
+//
+
+// closeWriteSession flushes and closes the specified |writeSession| and returns an error if anything failed.
+func closeWriteSession(ctx *sql.Context, engine *gms.Engine, databaseName string, writeSession writer.WriteSession) error {
```

Comment:
Good question... I got a hint on `WriteSession` being the right interface to use and this was just how I happened to get to it. It makes a ton of sense to use the WriteSession from the session, but I hit some snags trying that out yesterday. Most of the tests were passing, but I saw some strange behavior with changes across multiple databases. 

Instead, I refactored the Write/Update/Delete handlers into a common `processRowEvent` function so that all the WriteSession/TableWriter code was consolidated and identical for each row operation, and I changed it to batch edits per table, instead of each individual row. This should improve the performance slightly, but also make it easier for us to refactor in the future to move to the session managed WriteSession and debug the multi-db issues I was seeing.